Datadog trace flushing/export

ericmustin opened the pull request:
**Description:** This PR adds flushing+export of traces and trace-related statistics to the `datadogexporter`, as well as some very minor changes to the translation of internal traces into Datadog format. It represents the second of two PRs for the work contained in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1203. It builds on top of current master branch, and follows up to the work [done here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1208).

The final PR explicitly enabling The Datadog exporter will follow, and will allow users to export traces to Datadog's API Intake. 

This PR Split was requested by @tigrannajaryan and hopefully should make code review a bit less cumbersome. However if there are any questions or changes to the PR format needed, please let me know.

**Testing:** There are unit tests for the different methods and helper methods within the export code.
 
**Documentation:**  Appropriate usage, including best practices for which processors to also enable, has been documented in the README, `testdata/config.yaml` and `example/config.yaml` samples.

**Notes**: This PR includes a trace exporter for non-windows environments only (metrics are fine in windows, just traces that are the issue), due to reasons explained in this pr https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1274 . tl;dr is our trace export code for windows env would rely on CGO for now, which is not permitted in the collector

ericmustin committed:
feat: add datadog trace translation helpers and tests

ericmustin committed:
tests: add more unit tests for code cov

ericmustin committed:
tests: additional unit tests for codecov

ericmustin committed:
Merge branch 'master' into trace_translation
merge master

ericmustin committed:
dont modify go.sum since ddog exporter is not enabled

ericmustin committed:
test trace state and nil cases better

ericmustin committed:
Merge branch 'master' into trace_translation
merge upstream

ericmustin committed:
Merge branch 'master' into trace_translation
merge upstream

ericmustin committed:
exporter/datadog: update version and implement feedback

ericmustin committed:
Merge branch 'trace_translation' into datadog_trace_export
merge trace_translation

ericmustin committed:
exporters/datadog: add trace export and flush, update translation, add http utils

ericmustin committed:
exporters/datadog: force skipverify to always be false

ericmustin committed:
merge upstream

ericmustin committed:
remove unneeded methods and ensure utc time

ericmustin committed:
try not to build on windows

ericmustin committed:
add comment to not build on windows to all datadog files

mx-psi commented on exporter/datadogexporter/factory.go:
```diff
@@ -100,3 +104,28 @@ func createMetricsExporter(
 	)
 }
+
+// createMetricsExporter creates a trace exporter based on this config.
```
```suggestion
// createTraceExporter creates a trace exporter based on this config.
```

mx-psi commented on exporter/datadogexporter/README.md:
```diff
@@ -22,3 +22,37 @@ datadog:
 
 See the sample configuration file under the `example` folder for other available options.
+
+*IMPORTANT FOR TRACE EXPORT*: This exporter assumes that a pipeline which includes the datadog exporter also includes a [batch processor](https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/batchprocessor) configured with a `timeout` setting of `10s`(10 seconds). A batch representing 10 seconds of traces is a constraint of Datadog's API Intake for Trace Related Statist ...
```
I would make this a subsection.

ericmustin replied:
Makes sense, updated

mx-psi reviewed (changes requested):
(Reviewing on behalf of Datadog). We want to have support for Windows (even if it is limited) so we shouldn't add the `build !windows` build tag everywhere. I think it should be fine for Windows, it's just that a dependency uses Cgo so we need to add `gcc` to the CI.

edit: Since the project does not allow for the use of CGo we are disabling the traces exporter on Windows temporarily while we work around this limitation.

ericmustin committed:
exporters/datadog: only build metrics related code on windows

ericmustin committed:
exporters/datadog: return early on windows from trace factory

ericmustin committed:
exporters/datadog: fix language

ericmustin committed:
exporters/datadog: linting

ericmustin committed:
exporters/datadog:add stubbed windows trace exporter

ericmustin committed:
exporters/datadog: stub out trace export funcs for windows

ericmustin committed:
exporters/datadog: try to get stubs to work

ericmustin committed:
exporters/datadog: linting

mx-psi commented on exporter/datadogexporter/traces_exporter_windows.go:
```diff
@@ -0,0 +1,38 @@
+type traceExporterStub struct {}
+
+func newTraceExporter(logger *zap.Logger, cfg *Config) (*traceExporterStub, error) {
+	return &traceExporterStub{}, errors.New("datadog trace export is currently not supported on Windows")
```
Why not just return `nil` instead of the stub?

mx-psi replied:
You would have to change where you define things a bit so consider this as solved.

ericmustin replied:
There were some compilation issues in `factory.go` because we call some methods on the `traceExporter` return value, such as [pushTraceData](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1266/files#diff-c9e9f39ffda08a2af74c0d54d37f21e14cdb95c4136cdbe238de6be2bfe31389R133). So just returning stubs here made all those issues go away

mx-psi reviewed (approved):
You should merge master and resolve the conflict, I also left a nit/question. Otherwise LGTM!

ericmustin committed:
exporter/datadog: merge upstream master and update Readme

ericmustin committed:
exporters/datadog: merge upstream

ericmustin committed:
exporter/datadog: merge upstream again

ericmustin committed:
exporter/datadog: tests added some new things to go.mod

ericmustin committed:
Merge branch 'master' into datadog_trace_export
merge upstream

ericmustin committed:
exporter/datadog: increase timeout

jpkrohling commented on exporter/datadogexporter/traces_exporter_test.go:
```diff
@@ -0,0 +1,303 @@
+	ctx := context.Background()
+	errConsume := exporter.ConsumeTraces(ctx, td)
+	assert.NoError(t, errConsume)
+	exporter.Shutdown(context.Background())
```
I would move this closer to the `createTraceExporter`, with a defer.

jpkrohling commented on exporter/datadogexporter/traces_exporter_test.go:
```diff
@@ -0,0 +1,303 @@
+
+	// ensure that the protobuf serialized traces payload contains HostName Env and Traces
+	// ensure that the json gzipped stats payload contains HostName Env and Stats
+	got := testTraceExporterHelper(internaldata.OCToTraceData(td), t)
```
I think we really shouldn't be using `internaldata` anymore.

ericmustin replied:
Gotcha, no worries, just was using it to generate some mock data like the other vendor exporter tests, but can update these mocks when i get a sec to use the preferred format.

ericmustin replied:
So i spent some time looking into this. This would be trivial to mock properly without the OC helper funcs, using something like: `testdata.GenerateTraceDataTwoSpansSameResourceOneDifferent()` except that, If i understand correctly, `"go.opentelemetry.io/collector/internal/data/testdata"` is not accessible for the contrib repo since it's an internal package to opentelemetry-collector.

So I'm quite stuck here, as I don't see any other examples of how to mock this info in tests without a helper func, no other exporters have to do this in contrib without relying on OC helers, and the exporter's that handle it in core have those internal helper funcs. Wouldn't it just be easier if `go.opentelemetry.io/collector/internal/data/testdata` were accessible to contrib tests?

jpkrohling replied:
Moving to internal is something relatively new. What I've been doing is to create a helper func close to where it's used, like this:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a4b57fc84b100c3a5e852360a26be18be5cdfecb/processor/groupbytraceprocessor/processor_test.go#L771-L788

ericmustin replied:
makes sense, i've gone ahead and used a helper along these lines and removed the OC stuff

mx-psi commented on exporter/datadogexporter/factory.go:
```diff
@@ -100,3 +104,32 @@ func createMetricsExporter(
+		return nil, err
+	}
+
+	return exporterhelper.NewTraceExporter(
```
This sounds like a good solution to me *if* we can avoid the check when we only have the metrics part of the exporter enabled, otherwise I would just go with a loud warning

ericmustin replied:
@jpkrohling I agree that this is a good approach however don't really see a way to check for processors on the `component.Host` as far as i can tell, https://github.com/open-telemetry/opentelemetry-collector/blob/96767632d2208306de42bd3e5a531114e4b7cf02/component/component.go#L61-L94 . 

So I think we'd be blocked on adding the `Start` method discussed due to that that limitation in core. 

Broadly speaking i think we're Ok without a hard requirement on a 10s batch processor being set, for now. I agree with your comments here, https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1266#discussion_r508344022 that this is not a wonderful user experience, but it doesn't necessarily break anything if not followed, it just leads to potential metric inaccuracy for some metrics related to the traces. The traces themselves are fine. We'd ideally like to just get this in the hands of users and see their feedback before overthinking the setup restrictions/details too much.

jpkrohling replied:
Sounds good to me.

ericmustin committed:
Merge branch 'master' into datadog_trace_export
merge upstream

ericmustin commented on exporter/datadogexporter/stats.go:
```diff
@@ -0,0 +1,79 @@
+			// TODO: This is all hardcoded to assume 10s buckets for now
+
+			// Aggregate the span to a bucket by rounding its end timestamp to the closest bucket ts.
+			// E.g., for buckets of size 10, a span ends on 36 should be aggregated to the second bucket
```
So, some of this got copy/pasta'd in from another repo. This comment is confusing and i'm changing it to reflect the current state of the code.  TL;DR ideally we'd keep multiple buckets in a buffer with a broad time frame, and only flush the absolute oldest bucket every 10s. This would allow traces that get reported late to have a higher chance of ending up in their appropriate bucket. This is how our vendor specific datadog-agent handles things, but it's big ask to port that logic over as it's very tightly coupled with other code in the agent. So this simplified approach is what we've chosen for now to at least ensure we accurately count hits, even if the hits are bucketed off by +- 10s or so.

ericmustin commented on exporter/datadogexporter/stats.go:
```diff
@@ -0,0 +1,79 @@
+			// E.g., for buckets of size 10, a span ends on 36 should be aggregated to the second bucket
+			// with bucketTS 30 (36 - 36 % 10). Create a new bucket if needed.
+			// spanEnd := span.Start + span.Duration
+			// bucketTS := spanEnd - (spanEnd % statsBucketDuration)
```
removed

ericmustin committed:
exporter/datadog: implement feedback'

ericmustin committed:
merge upstream

ericmustin commented on exporter/datadogexporter/README.md:
```diff
@@ -25,4 +26,45 @@ See the sample configuration file under the `example` folder for other available
+### **Important Pipeline Setup Details** 
+
+This exporter assumes a pipeline using the datadog exporter also includes a [batch processor](https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/batchprocessor) configured with the following: 
+  - a `timeout` setting of `10s`(10 seconds). 
```
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1266/files#r508537804 . I think `send_batch_size` is good to keep in mind, i'm fine with that default for now. I think ultimately we'll have to iterate here around what best practice we should suggest to users, and as long as there's a default here that would protect users from blowing up mem/cpu perf from a buffer that receives too many traces at once, i think its a fine place to start.

ericmustin commented on exporter/datadogexporter/stats.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+const (
+	statsBucketDuration int64 = 1e10 // 10 seconds
```
keeping it explicitly as int64 was just easier since it has to match `pushTime` here https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1266/files#diff-9dabb193030fe7865f54b404f3e72c4e8d3289b1e5ab3c8fc6c076e2253416cbR33 which is the same type  (it's the unix nano timestamp).

jpkrohling replied:
Alright, but you could still have something like `int64(10 * time.Second)`, so that you can get rid of the comment while still easily telling that this is 10s.

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+			break
+		}
+
+		time.Sleep(traceEdgeRetryInterval)
```
Yup that makes sense. To be honest exponential backoff is probably a bit overkill here, in other implementations this is really only supposed to retry a few times at most, so I just bumped it to a 10s retry, and we'll be aware of keeping max retries relatively low, which I think should be fine to start

jpkrohling replied:
How does that relate to the timeout? Is the timeout across all attempts, or is it per attempt? If it's across all of them, you'll now be timing out before retrying.

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payload.Bytes))
+	req = req.WithContext(ctx)
+
+	if err != nil {
```
👍 makes sense, updated

ericmustin commented on exporter/datadogexporter/traces_exporter.go:
```diff
@@ -0,0 +1,119 @@
+
+	// this is for generating metrics like hits, errors, and latency, it uses a separate endpoint than Traces
+	stats := ComputeAPMStats(ddTracePayload, pushTime)
+	errStats := exp.edgeConnection.SendStats(context.Background(), stats, maxRetries)
```
ah yup, good catch. updated

ericmustin commented on exporter/datadogexporter/factory_test.go:
```diff
@@ -89,3 +91,32 @@ func TestCreateAPIMetricsExporter(t *testing.T) {
+	logger := zap.NewNop()
+
+	factories, err := componenttest.ExampleComponents()
+	assert.NoError(t, err)
```
updated

ericmustin commented on exporter/datadogexporter/factory_test.go:
```diff
@@ -89,3 +91,32 @@ func TestCreateAPIMetricsExporter(t *testing.T) {
+		cfg.Exporters["datadog/api"],
+	)
+
+	assert.Nil(t, err)
```
👍 updated

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+const (
+	traceEdgeTimout        time.Duration = 10 * time.Second
```
good catch 😅

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+func (con *traceEdgeConnection) SendTraces(ctx context.Context, trace *pb.TracePayload, maxRetries int) error {
+	binary, marshallErr := proto.Marshal(trace)
+	if marshallErr != nil {
+		return fmt.Errorf("failed to serialize trace payload to protobuf: %v", marshallErr)
```
👍

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+		return fmt.Errorf("no traces in payload")
+	}
+
+	fmt.Printf("Sending trace %d\n", trace.Traces[0].TraceID)
```
👍

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+
+		time.Sleep(traceEdgeRetryInterval)
+	}
+	return fmt.Errorf("failed to send trace payload to trace edge: %v", sendErr)
```
👍

ericmustin commented on exporter/datadogexporter/trace_connection.go:
```diff
@@ -0,0 +1,183 @@
+	var b bytes.Buffer
+	err := stats.EncodePayload(&b, sts)
+	if err != nil {
+		return fmt.Errorf("failed to encode stats payload: %v", err)
```
👍

ericmustin commented on exporter/datadogexporter/traces_exporter_test.go:
```diff
@@ -0,0 +1,303 @@
+		got = append(got, data...)
+
+		if contentType == "application/x-protobuf" {
+			var traceData pb.TracePayload
```
updated  still a bit clunky but more readable now at least

ericmustin commented on exporter/datadogexporter/traces_exporter_test.go:
```diff
@@ -0,0 +1,303 @@
+
+		contentType := req.Header.Get("Content-Type")
+
+		data := []string{req.Header.Get("Content-Type")}
```
👍

ericmustin committed:
exporter/datadog: use new format of exporter

jpkrohling commented on exporter/datadogexporter/traces_exporter_test.go:
```diff
@@ -0,0 +1,315 @@
+		return traces
+	}())
+
+	assert.Nil(t, err)
```
If you are updating this PR for some other reason, update also calls like this to use `NoError`.

ericmustin replied:
👍 Updated

jpkrohling reviewed (commented):
LGTM. There are a few things that can be improved in the future, but I think it's sane. One thing to clarify is on the usage of OC for the test data: for new code, I think we should really try to avoid using OC.

ericmustin committed:
exporter/datadog: simplify mocks, add proper nil tests, use more readable timing methods

jpkrohling reviewed (approved):
LGTM.

ericmustin commented:
@jpkrohling I've updated the helpers/mocks to not require OC helpers, using the approach you suggested. I think this is much cleaner, ty for suggestion.

also re: timeouts, the timeout is per retry so, it should be ok?

and last, cleaned up the other bits you highlights around `.noError` and using better `time` conventions for readability.
